### ISSUE TITLE
Armv7 build

### ISSUE THREAD

**Issue by mladenp87** [jacobalberty/firebird-docker#67]
Can we expect Armv7 support ? :) 
Raspberry pi users would be grateful :)

Best regards!

**Comment by jacobalberty** [comment]
That is planned, I was not able to build the 4.0 betas on arm though, I'm trying to find more information on if it is still buildable on it. I will give the 3.0 stuff a try soon and see if I can get it building there. I did just revamp the build system to make that doable though.

**Comment by mladenp87** [comment]
Thank you for the reply. Looking forward to it.

**Comment by mariuz** [comment]
4.0 final should be ok on arm at least on ARM64 and 32 (see the android builds)

https://firebirdsql.org/en/firebird-4-0-0/

Could you also push images for arm64 ?

https://docs.docker.com/desktop/multi-arch/#build-and-run-multi-architecture-images

**Comment by jacobalberty** [comment]
The build script is set up to do multi arch I just didn't enable arm for firebird.  I do not (yet) have an arm 64 device set up to test functionality on. But I am testing build on 32 bit now now to make sure there are no issues, I'll try pushing 64 bit with 32 bit once i confirm at least 32 bit builds without issue. I have a  pi 4b arriving tomorrow that i will set up 64 bit linux on to verify functionality.

**Comment by jacobalberty** [comment]
32 bit arm built with minimal issues.  64 bit arm however is giving me some trouble.

```
#11 212.4
#11 212.4 This script, last modified 2010-08-21, has failed to recognize
#11 212.4 the operating system you are using. It is advised that you
#11 212.4 download the most up to date version of the config scripts from
#11 212.4
#11 212.4   http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
#11 212.4 and
#11 212.4   http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD
#11 212.4                                                                                                                                                               #11 212.4 If the version you run (builds/make.new/config/config.guess) is already up to date, please
#11 212.4 send the following data and any information you think might be
#11 212.4 pertinent to <config-patches@gnu.org> in order to provide the needed
#11 212.4 information to handle your system.
```

I'll try building on real hardware this weekend and see if its just a bug in buildx/qemu or what.

**Comment by jacobalberty** [comment]
Post office didnt get my raspberry pi here by expected date. I'm assuming it will be here tomorrow so I'll hopefully have a better idea tomorrow evening of whats going on.

I see on the mailing list it just builds under ubuntu 20.04 on raspberry pi 4 hardware without issue. I'll try building under raspbian natively as well as under docker on the raspbian system running on native hardware. I suspect it is a qemu bug preventing autoconf from properly detecting the system. It could also be a bug in debian on arm64 though.


There was more debugging information in the ./configure output that is possibly relevant. But I want to narrow down the issue to being a debian on arm64 or qemu bug first before I included it.

Right now on my x86_64 docker system i'm using the following commands to build:
`docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`
`docker buildx create --use`
those two set up the buildx system then I'm trying to build the arm64 version with:
`docker buildx build --progress plain --platform linux/arm64/v8`

qemu unfortunately does have bugs relating to arm64, so I suspect thats where the issue lies.

**Comment by jacobalberty** [comment]
Ok I was able to install debian buster 64 bit on a raspberry pi 4b with 8GB of ram. It fails at the exact same space. So this appears to be an issue with firebird on debian on arm64 and not qemu or docker specific
Full error message is at end of this comment. It looks simple enough it just appears to not recognize the system for some reason.  This may need to be brought over to the official firebird mailing list though as apparently this builds on ubuntu arm64 but not on debian arm64. Maybe there's just a utility it wants  to identify the system properly that I can just install though?

```
quake@rpi4-20210705:~/firebird$ ./configure
checking whether make sets $(MAKE)... no
checking build system type... builds/make.new/config/config.guess: unable to guess system type

This script, last modified 2010-08-21, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
and
  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD

If the version you run (builds/make.new/config/config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2010-08-21

uname -m = aarch64
uname -r = 5.10.0-0.bpo.7-arm64
uname -s = Linux
uname -v = #1 SMP Debian 5.10.40-1~bpo10+1 (2021-06-04)

/usr/bin/uname -p = unknown
/bin/uname -X     =

hostinfo               =
/bin/universe          =
/usr/bin/arch -k       =
/bin/arch              = aarch64
/usr/bin/oslevel       =
/usr/convex/getsysinfo =

UNAME_MACHINE = aarch64
UNAME_RELEASE = 5.10.0-0.bpo.7-arm64
UNAME_SYSTEM  = Linux
UNAME_VERSION = #1 SMP Debian 5.10.40-1~bpo10+1 (2021-06-04)
configure: error: cannot guess build type; you must specify one
```

**Comment by jacobalberty** [comment]
If I apply the attached patch and add `--build=aarch64-linux-gnu` to the configure line it builds.
Adding a patch like that is no problem, thats what the pre_build and post_build directories are for.
The only problem I have now is there's no easy way for me to change the configure line without adding arm64 specific checks in build.sh. I would prefer to find a solution to correct the issue with detecting the platform before I do that. The mailing list mentions it building fine in ubuntu arm64 so I need to fire up ubuntu arm64 and play around there see if maybe I can find whats different between ubuntu and debian on arm64.

I really don't believe the solution is applying patches and platform specific checks. I believe there's a cleaner solution possibly in supplying one of those commands that the error log shows as empty output

[aarch64.diff.gz](https://github.com/jacobalberty/firebird-docker/files/6772623/aarch64.diff.gz)

**Comment by jacobalberty** [comment]
Ok, so ubuntu has the same issue

```
root@ubuntu:/home/firebird# ./configure
checking whether make sets $(MAKE)... no
checking build system type... builds/make.new/config/config.guess: unable to guess system type

This script, last modified 2010-08-21, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
and
  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD

If the version you run (builds/make.new/config/config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2010-08-21

uname -m = aarch64
uname -r = 5.4.0-1038-raspi
uname -s = Linux
uname -v = #41-Ubuntu SMP PREEMPT Thu Jun 17 14:14:11 UTC 2021

/usr/bin/uname -p = aarch64
/bin/uname -X     =

hostinfo               =
/bin/universe          =
/usr/bin/arch -k       =
/bin/arch              = aarch64
/usr/bin/oslevel       =
/usr/convex/getsysinfo =

UNAME_MACHINE = aarch64
UNAME_RELEASE = 5.4.0-1038-raspi
UNAME_SYSTEM  = Linux
UNAME_VERSION = #41-Ubuntu SMP PREEMPT Thu Jun 17 14:14:11 UTC 2021
configure: error: cannot guess build type; you must specify one
root@ubuntu:/home/firebird#
```

**Comment by jacobalberty** [comment]
I ended up just pulling a new copy of config.guess from the gnu project and it seems to be building correctly now. 
I have github actions building it now it should be done in about 2 hours if anyone wants to test it on real hardware. I have reverted my raspberry pi back to stock 32 bit raspios  for its more permanent use so I can not test the 64 bit images. I will give 32 bit  arm images a test when they finish building though.
